src/mds/CInode.h

```diff
@@ -67,17 +68,67 @@ struct cinode_lock_info_t {
  */
 class InodeStoreBase {
 public:
-  typedef inode_t<mempool::mds_co::pool_allocator> mempool_inode;
-  typedef old_inode_t<mempool::mds_co::pool_allocator> mempool_old_inode;
-  typedef mempool::mds_co::compact_map<snapid_t, mempool_old_inode> mempool_old_inode_map;
-  typedef xattr_map<mempool::mds_co::pool_allocator> mempool_xattr_map; // FIXME bufferptr not in mempool
+  using mempool_inode = inode_t<mempool::mds_co::pool_allocator>;
+  using inode_ptr = std::shared_ptr<mempool_inode>;
+  using inode_const_ptr = std::shared_ptr<const mempool_inode>;
+
+  template <typename ...Args>
+  static inode_ptr allocate_inode(Args && ...args) {
+    static mempool::mds_co::pool_allocator<mempool_inode> allocator;
+    return std::allocate_shared<mempool_inode>(allocator, std::forward<Args>(args)...);
+  }
+  static inode_ptr const_inode_cast(const inode_const_ptr& ptr) {
```

Comment:
This method (and the others like it) bothers me. You're dropping the `const`-ness of the ptr. I've looked through the code and the main reason for this looks to be the various `InodeStoreBase::decode_*` methods? Can't we invert the logic so that the `decode_*` method allocates the new inode and assigns it to the `InodeStoreBase::inode` to avoid this magic.

Reply:
There's also `get_inode()` which provides access to the old-style access and allows breaking the `const`-ness.

Looking at this code it's hard for me to say where `*InodeStoreBase::inode` can be changed. I understand that's harder to fix perhaps but it could use comments. Is that worth fixing in a future refactor?

Reply:
const_inode_cast is removed.  _get_inode() is still there, it's only used for new inode.